src/libmeasurement_kit/engine.cpp

```diff
@@ -104,17 +100,22 @@ Task::Task(nlohmann::json &&settings) {
         pimpl_->running = true;
         barrier.set_value();
         static Semaphore semaphore;
-        semaphore.acquire(); // block until a previous task has finished running
-        task_run(pimpl_.get(), settings);
+        task_run(pimpl_.get(), settings, [&]() {
+            semaphore.acquire(); // block until a previous task has finished
```

Comment:
my only comment on this is that there's a lot of complexity in the concurrency here. semaphores, barriers, lambdas, waiters. 

a bit more comments explaining the state machine, or some tests to try to limit the potential of race conditions entering this would be useful